pennylane/ops/op_math/controlled_class.py

```diff
@@ -0,0 +1,380 @@
+    >>> op
+    CRX(1.234, wires=[0, 1])
+
+    The operation has both standard :class:`~.operation.Operator` properties
+    and ``Controlled`` specific properties:
+
+    >>> op.base
+    RX(1.234, wires=[1])
+    >>> op.data
+    [1.234]
+    >>> op.wires
+    <Wires = [0, 1]>
+    >>> op.control_wires
+    <Wires = [0]>
+    >>> op.target_wires
+    <Wires = [1]>
+
+    Control values are lists of booleans, indicating whether or not to control on the
+    ``0==False`` value or the ``1==True`` wire.
+
+    >>> op.control_values
+    [0]
+
+    Representations for an operator are available if the base class defines them.
+    Sparse matrices are available if the base class defines either a sparse matrix
+    or only a dense matrix.
+
+    >>> np.set_printoptions(precision=4) # easier to read the matrix
+    >>> qml.matrix(op)
+    array([[0.8156+0.j    , 0.    -0.5786j, 0.    +0.j    , 0.    +0.j    ],
+           [0.    -0.5786j, 0.8156+0.j    , 0.    +0.j    , 0.    +0.j    ],
+           [0.    +0.j    , 0.    +0.j    , 1.    +0.j    , 0.    +0.j    ],
+           [0.    +0.j    , 0.    +0.j    , 0.    +0.j    , 1.    +0.j    ]])
+    >>> qml.eigvals(op)
+    array([1.    +0.j    , 1.    +0.j    , 0.8156+0.5786j, 0.8156-0.5786j])
+    >>> qml.generator(op, format='observable')
+    (-0.5) [Projector0 X1]
+    >>> op.sparse_matrix()
+    <4x4 sparse matrix of type '<class 'numpy.complex128'>'
+                with 6 stored elements in Compressed Sparse Row format>
+
+    If the provided base matrix is an :class:`~.operation.Operation`, then the created
+    object will be of type :class:`~.ops.op_math.ControlledOp`. This class adds some additional methods
+    and properties to the basic :class:`~.ops.op_math.Controlled` class.
+
+    >>> type(op)
+    pennylane.ops.op_math.controlled_class.ControlledOp
+    >>> op.parameter_frequencies
+    [(0.5, 1.0)]
+
+    """
+
+    # pylint: disable=unused-argument
+    def __new__(cls, base, *_, **__):
+        """If base is an ``Operation``, then the a ``ControlledOp`` should be used instead."""
+        if isinstance(base, operation.Operation):
+            return object.__new__(ControlledOp)
+        return object.__new__(Controlled)
+
+    # pylint: disable=too-many-function-args
+    def __init__(
+        self, base, control_wires, control_values=None, work_wires=None, do_queue=True, id=None
+    ):
+        control_wires = Wires(control_wires)
+        work_wires = Wires([]) if work_wires is None else Wires(work_wires)
+
+        if control_values is None:
+            control_values = [True] * len(control_wires)
+        else:
+            if isinstance(control_values, str):
+                warnings.warn(
+                    "Specifying control values as a string is deprecated. Please use Sequence[Bool]",
+                    UserWarning,
+                )
+                control_values = [(x == "1") for x in control_values]
+
+            assert len(control_values) == len(
```

Comment:
Why are you using `assert`? Assert is commonly used for checking bugs, but these errors depend on the user input.

Reply:
Usually in PennyLane, we raise some kind of error with problematic input, but I'm now leaning toward using `assert` statements in this kind of case.

These statements are for checking unexpected behaviour.  In most use cases, they won't be triggered, especially if the user knows what they are doing. By using `assert`, I'm giving the user a choice not to run input validation.

Reply:
Exactly! As you mentioned, `assert` is generally used to check unexpected behaviour of the code. But these asserts depend on the values that the user inputs, thus raising an error should be preferred in these cases. Why do you want to give the user the choice of skipping input validation? To improve performance?

Reply:
Efficiency.

We can tend to do a lot of input validation in PennyLane, and that ends up with us having a lot of `isinstance` checks and other forms of validation.  I've been thinking that it would be a good idea to allow high-performance users to cut out the validation somehow, and using `assert` statements seemed like a good way of doing that.

But I've now further researched it, and it sounds like the internet discourages its use in input validation, so I'll change it back to `ValueError`.

Maybe we look into a custom global variable that allows power users to turn off validation?

Reply:
Yes, using a global variable sounds more reasonable.